packages/governance/src/votes/votes.cairo

```diff
@@ -1,40 +1,44 @@
 // SPDX-License-Identifier: MIT
-// OpenZeppelin Contracts for Cairo v0.17.0 (token/erc20/extensions/erc20_votes.cairo)
+// OpenZeppelin Contracts for Cairo v0.17.0 (governance/votes/votes.cairo)
 
-use core::hash::{HashStateTrait, HashStateExTrait};
-use core::poseidon::PoseidonTrait;
-use openzeppelin_utils::cryptography::snip12::{OffchainMessageHash, StructHash, SNIP12Metadata};
 use starknet::ContractAddress;
 
-/// # ERC20Votes Component
+
+/// # Votes Component
+///
+/// The Votes component provides a flexible system for tracking voting power and delegation
+/// that is currently implemented for ERC20 and ERC721 tokens. It allows accounts to delegate
+/// their voting power to a representative, who can then use the pooled voting power in
+/// governance decisions. Voting power must be delegated to be counted, and an account can
+/// delegate to itself if it wishes to vote directly.
 ///
-/// The ERC20Votes component tracks voting units from ERC20 balances, which are a measure of voting
-/// power that can be transferred, and provides a system of vote delegation, where an account can
-/// delegate its voting units to a sort of "representative" that will pool delegated voting units
-/// from different accounts and can then use it to vote in decisions. In fact, voting units MUST be
-/// delegated in order to count as actual votes, and an account has to delegate those votes to
-/// itself if it wishes to participate in decisions and does not have a trusted representative.
+/// This component offers a unified interface for voting mechanisms across ERC20 and ERC721
+/// token standards, with the potential to be extended to other token standards in the future.
+/// It's important to note that only one token implementation (either ERC20 or ERC721) should
+/// be used at a time to ensure consistent voting power calculations.
```

Comment:
This component is not pegged to ERC20 and ERC721 even when it provides implementation of `get_voting_units` for them. We may add an implementation for a component that is not even a token, or users may use it to provide their own implementation for their custom contracts (similar to how the Votes extension in solidity is not TokenVotes, even when can be used in tokens). I think we can rephrase this a bit because of that, and also change the name of the `TokenVotesTrait` trait to something more generic like `VotingUnitsTrait`.

I suggest something like this:

```
/// # Votes Component
///
/// The Votes component provides a flexible system for tracking and delegating voting power.
/// that is currently implemented for ERC20 and ERC721 tokens. An account can delegate
/// their voting power to a representative, that will pool delegated voting units from different
/// delegators and can then use it to vote in decisions. Voting power must be delegated to be counted,
/// and an account must delegate to itself if it wishes to vote directly without a trusted
/// representative.
///
/// When integrating the Votes component, the ´VotingUnitsTrait´ must be implemented to get the voting
/// units for a given account as a function of the implementing contract. For simplicity, this module
/// already provides two implementations for ERC20 and ERC721 tokens, which will work out of the box
/// if the respective components are integrated.
///
/// NOTE: ERC20 and ERC721 tokens implementing this component must call ´transfer_voting_units´
/// whenever a transfer, mint, or burn operation is performed. Hooks can be leveraged for this purpose,
/// as shown in the following ERC20 example:
```

And I think is worth adding the full example since we don't have a preset or a Wizard example for this:

```
/// NOTE: ERC20 and ERC721 tokens implementing this component must call ´transfer_voting_units´
/// whenever a transfer, mint, or burn operation is performed. Hooks can be leveraged for this purpose,
/// as shown in the following ERC20 example:
///
/// ```cairo
/// #[starknet::contract]
/// pub mod ERC20VotesPreset {
///     use openzeppelin_governance::votes::VotesComponent;
///     use openzeppelin_token::erc20::ERC20Component;
///     use openzeppelin_utils::cryptography::nonces::NoncesComponent;
///     use openzeppelin_utils::cryptography::snip12::SNIP12Metadata;
///     use starknet::ContractAddress;
///
///     component!(path: VotesComponent, storage: erc20_votes, event: ERC20VotesEvent);
///     component!(path: ERC20Component, storage: erc20, event: ERC20Event);
///     component!(path: NoncesComponent, storage: nonces, event: NoncesEvent);
///
///     #[abi(embed_v0)]
///     impl VotesImpl = VotesComponent::VotesImpl<ContractState>;
///     impl VotesInternalImpl = VotesComponent::InternalImpl<ContractState>;
///
///     #[abi(embed_v0)]
///     impl ERC20MixinImpl = ERC20Component::ERC20MixinImpl<ContractState>;
///     impl ERC20InternalImpl = ERC20Component::InternalImpl<ContractState>;
///
///     #[abi(embed_v0)]
///     impl NoncesImpl = NoncesComponent::NoncesImpl<ContractState>;
///
///     #[storage]
///     pub struct Storage {
///         #[substorage(v0)]
///         pub erc20_votes: VotesComponent::Storage,
///         #[substorage(v0)]
///         pub erc20: ERC20Component::Storage,
///         #[substorage(v0)]
///         pub nonces: NoncesComponent::Storage
///     }
///
///     #[event]
///     #[derive(Drop, starknet::Event)]
///     enum Event {
///         #[flat]
///         ERC20VotesEvent: VotesComponent::Event,
///         #[flat]
///         ERC20Event: ERC20Component::Event,
///         #[flat]
///         NoncesEvent: NoncesComponent::Event
///     }
///
///     pub impl SNIP12MetadataImpl of SNIP12Metadata {
///         fn name() -> felt252 {
///             'DAPP_NAME'
///         }
///         fn version() -> felt252 {
///             'DAPP_VERSION'
///         }
///     }
///
///     impl ERC20VotesHooksImpl<
///         TContractState,
///         impl Votes: VotesComponent::HasComponent<TContractState>,
///         impl HasComponent: ERC20Component::HasComponent<TContractState>,
///         +NoncesComponent::HasComponent<TContractState>,
///         +Drop<TContractState>
///     > of ERC20Component::ERC20HooksTrait<TContractState> {
///         fn after_update(
///             ref self: ERC20Component::ComponentState<TContractState>,
///             from: ContractAddress,
///             recipient: ContractAddress,
///             amount: u256
///         ) {
///             let mut votes_component = get_dep_component_mut!(ref self, Votes);
///             votes_component.transfer_voting_units(from, recipient, amount);
///         }
///     }
///
///     #[constructor]
///     fn constructor(ref self: ContractState) {
///         self.erc20.initializer("MyToken", "MTK");
///     }
/// }
```

Reply:
Agree with the description. At first I didn't include much references to `TokenVotesTrait`(now `VotingUnitsTrait`) because [we didn't want the use implementing this trait](https://github.com/OpenZeppelin/cairo-contracts/pull/1114#discussion_r1734468508), but I think if someone is trying to come up with a novel governance mechanism that includes other ways to track votes, this would be helpful for sure.
About including the code example I went ahead and added, but have a few questions:
1. Since there's no place such as `docs.rs` where people can go ahead and view autogenerated docs from the comments, how valuable is it?
2. I don't think we've added this for other components without presets or wizard in the past? e.g. `timelock`. So this might impact consisntency.

@andrew-fleming do you have thoughts here?

Reply:
TBH I probably wouldn't include a full contract example in the code doc description. We already have a contract written out in `Usage` on the governance docsite page. How about we provide a link to that instead?

Reply:
I'm not quite convinced myself about adding the example in the comment, and I'm also leaning toward providing the link, but some points in favor of doing it:

> Since there's no place such as docs.rs where people can go ahead and view autogenerated docs from the comments, how valuable is it?

The language server shows help tooltips generated from comments when people hover over a type or module. We also have a tool for generating docs from comments similar to Rust that we are not using yet but might start using at some point.

It feels more rust-like to have some code examples in comments since it can be even compiled and checked directly at some point, and they are certainly easier to find (from LSP tooltips). With this said, I think we can wait until we have a docs.rs similar and/or the LSP works faster before adding this.

My main concern, and the thing that I'm trying to bring to the table is that I think it should be easier for users to find how to use this "kind of complex" components, and while the Usage section in the doc-site is certainly very helpful, I think we should be actively looking for ways to make it even easier (like Wizard, comments, etc...)

Reply:
> The language server shows help tooltips generated from comments when people hover over a type or module. We also have a tool for generating docs from comments similar to Rust that we are not using yet but might start using at some point.

I very much look forward to when we can autogen the docs on the docsite

> It feels more rust-like to have some code examples in comments since it can be even compiled and checked directly at some point, and they are certainly easier to find (from LSP tooltips). ~With this said, I think we can wait until we have a docs.rs similar and/or the LSP works faster before adding this.~

Agreed and I'm not opposed

> My main concern, and the thing that I'm trying to bring to the table is that I think it should be easier for users to find how to use these "kind of complex" components, and while the Usage section in the doc-site is certainly very helpful, I think we should be actively looking for ways to make it even easier (like Wizard, comments, etc...)

As I remember in the past, the idea was to direct users to the docsite in regard to using certain features, while we provided more of a barebones approach with the in-code docs (to encourage users to go to the docsite). If we want to be even more "rusty" with in-code docs, I think it's important to discuss and define what our goals and expectations should be with both sets of docs—probably worth establishing in the style guidelines

That said, I'm convinced with having the hooks impl in the component description 👍 I'd still provide a link to the full contract example as well

Reply:
I also like the idea of having more and better in code docs, but once they are useful for autogenerating on the docsite, and we should keep some consistency in the meantime. So conclusion would be to just leave the hooks part and a link to the docs? Something like this?
```
/// ```cairo
/// #[starknet::contract]
/// pub mod ERC20VotesContract {
/// (...)
///     impl ERC20VotesHooksImpl<
///         TContractState,
///         impl Votes: VotesComponent::HasComponent<TContractState>,
///         impl HasComponent: ERC20Component::HasComponent<TContractState>,
///         +NoncesComponent::HasComponent<TContractState>,
///         +Drop<TContractState>
///     > of ERC20Component::ERC20HooksTrait<TContractState> {
///         fn after_update(
///             ref self: ERC20Component::ComponentState<TContractState>,
///             from: ContractAddress,
///             recipient: ContractAddress,
///             amount: u256
///         ) {
///             let mut votes_component = get_dep_component_mut!(ref self, Votes);
///             votes_component.transfer_voting_units(from, recipient, amount);
///         }
///     }
/// }
```

/// See [the documentation]
/// (https://docs.openzeppelin.com/contracts-cairo/0.17.0/governance.html#usage_2)
/// for more details.

Reply:
We decided to just go with the reference to the usage section